x-pack/plugins/fleet/server/routes/security.ts

```diff
@@ -165,22 +93,199 @@ export async function getAuthzFromRequest(req: KibanaRequest): Promise<FleetAuth
+      if (!authz.fleet[key as keyof FleetAuthz['fleet']]) {
+        missingAuthz.push(`fleet.${key}`);
+      }
+    }
+  }
+
+  if (fleetAuthz && typeof fleetAuthz.integrations !== 'undefined') {
+    for (const key of fleetAuthz.integrations) {
+      if (!authz.integrations[key as keyof FleetAuthz['integrations']]) {
+        missingAuthz.push(`integrations.${key}`);
+      }
+    }
+  }
+
+  if (missingAuthz.length > 0) {
+    return false;
+  }
+
+  return true;
+}
+
+interface FleetAuthzRouteConfig {
+  fleet?: Array<keyof FleetAuthz['fleet']>;
+  integrations?: Array<keyof FleetAuthz['integrations']>;
 }
 
-export const RouterWrappers: RouterWrappersSetup = {
-  require: {
-    superuser: (router) => {
-      return makeRouterEnforcingSuperuser(router);
+type FleetAuthzRouteRegistrar<
+  Method extends RouteMethod,
+  Context extends RequestHandlerContext = RequestHandlerContext
+> = <P, Q, B>(
+  route: FleetRouteConfig<P, Q, B, Method>,
+  handler: RequestHandler<P, Q, B, Context, Method>
+) => void;
+
+type FleetRouteConfig<P, Q, B, Method extends RouteMethod> = RouteConfig<P, Q, B, Method> & {
+  fleetAuthz?: FleetAuthzRouteConfig;
+  fleetRequireSuperuser?: boolean;
+  // TODO temporary required while agents call Fleet setup
+  fleetAllowFleetSetupPrivilege?: boolean;
+};
+
+// Fleet router that allow to add required access when registering route
+export interface FleetAuthzRouter<
+  TContext extends FleetRequestHandlerContext = FleetRequestHandlerContext
+> extends IRouter<TContext> {
+  get: FleetAuthzRouteRegistrar<'get', TContext>;
+  delete: FleetAuthzRouteRegistrar<'delete', TContext>;
+  post: FleetAuthzRouteRegistrar<'post', TContext>;
+  put: FleetAuthzRouteRegistrar<'put', TContext>;
+  patch: FleetAuthzRouteRegistrar<'patch', TContext>;
+}
+
+function shouldHandlePostAuthRequest(req: KibanaRequest) {
+  return req.route.path.match(/^\/api\/fleet/);
+}
+
+export function makeRouterWithFleetAuthz<TContext extends FleetRequestHandlerContext>(
+  router: IRouter<TContext>
+): { router: FleetAuthzRouter<TContext>; onPostAuthHandler: OnPostAuthHandler } {
+  const authzMap = new Map<
+    string,
+    {
+      fleetAuthz?: FleetAuthzRouteConfig;
+      fleetAllowFleetSetupPrivilege?: boolean;
+      fleetRequireSuperuser?: boolean;
+    }
+  >();
+
+  function routeKey(routeOptions: { path: string; method: string }) {
+    return `${routeOptions.method}:${routeOptions.path}`;
+  }
+
+  function addRouteAuthz(
+    routeOptions: { path: string; method: string },
+    authConfig: {
+      fleetAuthz?: FleetAuthzRouteConfig;
+      fleetAllowFleetSetupPrivilege?: boolean;
+      fleetRequireSuperuser?: boolean;
+    }
+  ) {
+    authzMap.set(routeKey(routeOptions), authConfig);
+  }
+
+  const fleetAuthzOnPostAuthHandler: OnPostAuthHandler = async (req, res, toolkit) => {
+    if (!shouldHandlePostAuthRequest(req)) {
+      return toolkit.next();
+    }
+
+    if (!checkSecurityEnabled()) {
+      return res.forbidden();
+    }
```

Comment:
Will this break the `check-permissions` API?

Reply:
@nchaulet this is my only pending question before approving this

Reply:
@joshdover as the check permissions API do not require any fleet role we are not going through that code for this route, so it should change nothing for the check permissions API.